### PR TITLE
[SPARK-15681][core] allow lowercase or mixed case log level string when calling sc.setLogLevel

### PR DESCRIPTION
**Pull request by xwu0226**
## What changes were proposed in this pull request?

Currently `SparkContext API setLogLevel(level: String)`can not handle lower case or mixed case input string. But `org.apache.log4j.Level.toLevel` can take lowercase or mixed case. 

This PR is to allow case-insensitive user input for the log level.
## How was this patch tested?

A unit testcase is added.

### CODE REVIEW

**Comment by xwu0226** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 1]
```diff
@@ -357,13 +359,22 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    */
   def setLogLevel(logLevel: String) {
     val validLevels = Seq("ALL", "DEBUG", "ERROR", "FATAL", "INFO", "OFF", "TRACE", "WARN")
-    if (!validLevels.contains(logLevel)) {
+    // let's allow lowcase or mixed case too
+    if (!validLevels.contains(logLevel.toUpperCase)) {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: Can `validLevels` just become a `Set` in a `SparkContext` `object`? 
You can call `toUpperCase(Locale.ENGLISH)`, once, and save it in a val, and use the same value when checking the value and passing it to log4j, just for tidiness.

I don't think the `resetLogLevel` method is worth adding.

Comment (xwu0226) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: @srowen Thanks!  The `toUpperCase` is against the user-input `logLevel`, then check it against the `validLevels` right?  `validLevels` already contain a seq of upper case values. 

But i will change `toUpperCase` to `toUpperCase(Locale.ENGLISH)`
And i will remove the `resetLogLevel`.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 2]
```diff
@@ -357,7 +357,8 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    */
   def setLogLevel(logLevel: String) {
     val validLevels = Seq("ALL", "DEBUG", "ERROR", "FATAL", "INFO", "OFF", "TRACE", "WARN")
-    if (!validLevels.contains(logLevel)) {
+    // let's allow lowcase or mixed case too
+    if (!validLevels.contains(logLevel.toUpperCase(Locale.ENGLISH))) {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: You could use `require` here even.
My suggestion was to use a `Set` above, and, to make it a static member of an object. However, maybe that's really trivial since this method is called so infrequently.
I do think it's best to pass the value to log4j that is validated here -- the uppercase version.

Comment (xwu0226) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: @srowen Thanks! Oh. I see what you mean. I can move the validLevels as a set as a static member. And I can also store the `logLevel.toUpperCase` to a val and pass it to log4j. The reason I was passing `logLevel` as is is because `org.apache.log4j.Level.toLevel(logLevel)` also converts it to upper case internally.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 3]
```diff
@@ -2179,6 +2178,7 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
  * various Spark features.
  */
 object SparkContext extends Logging {
+  private val validLevels = Set("ALL", "DEBUG", "ERROR", "FATAL", "INFO", "OFF", "TRACE", "WARN")
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: Pretty minor, but I might call this `VALID_LOG_LEVELS`

File: core/src/test/scala/org/apache/spark/SparkContextSuite.scala [hunk 4]
```diff
@@ -363,4 +363,18 @@ class SparkContextSuite extends SparkFunSuite with LocalSparkContext {
...
+    sc = new SparkContext(new SparkConf().setAppName("test").setMaster("local"))
+    val originalLevel = org.apache.log4j.Logger.getRootLogger().getLevel
+    try {
+      sc.setLogLevel("debug")
+      assert(org.apache.log4j.Logger.getRootLogger().getLevel === org.apache.log4j.Level.DEBUG)
+      sc.setLogLevel("INfo")
+      assert(org.apache.log4j.Logger.getRootLogger().getLevel === org.apache.log4j.Level.INFO)
+    } finally {
+      sc.setLogLevel(originalLevel.toString)
+      assert(org.apache.log4j.Logger.getRootLogger().getLevel === originalLevel)
```
Comment (srowen) [inline core/src/test/scala/org/apache/spark/SparkContextSuite.scala]: You need to stop the context.

File: repl/scala-2.11/src/test/scala/org/apache/spark/repl/ReplSuite.scala [hunk 5]
```diff
@@ -21,6 +21,7 @@ import java.io._
 import java.net.URLClassLoader
 
 import scala.collection.mutable.ArrayBuffer
+
```
Comment (srowen) [inline repl/scala-2.11/src/test/scala/org/apache/spark/repl/ReplSuite.scala]: I'd undo this change

**Comment by xwu0226** [comment]
@srowen Thank you! I updated the code upon your comments.

**Comment by srowen** [comment]
LGTM

**Comment by srowen** [comment]
Jenkins retest this please

**Comment by xwu0226** [comment]
@srowen Thanks!

**Comment by vanzin** [comment]
LGTM; merging to master / 2.0.
